src/main/java/com/datapipe/jenkins/vault/credentials/AbstractVaultTokenCredentialWithExpiration.java

```diff
@@ -0,0 +1,75 @@
+package com.datapipe.jenkins.vault.credentials;
+
+import com.bettercloud.vault.Vault;
+import com.bettercloud.vault.VaultConfig;
+import com.bettercloud.vault.VaultException;
+import com.cloudbees.plugins.credentials.CredentialsScope;
+import java.util.Calendar;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class AbstractVaultTokenCredentialWithExpiration
+    extends AbstractVaultTokenCredential {
+
+    private final static Logger LOGGER = Logger
+        .getLogger(AbstractVaultTokenCredentialWithExpiration.class.getName());
+
+    private Calendar tokenExpiry;
+    private String currentClientToken;
+
+    protected AbstractVaultTokenCredentialWithExpiration(CredentialsScope scope, String id,
+        String description) {
+        super(scope, id, description);
+    }
+
+    protected abstract String getToken(Vault vault);
+
+    @Override
+    public Vault authorizeWithVault(VaultConfig config) {
+        Vault vault = getVault(config);
+        if (tokenExpired()) {
+            currentClientToken = getToken(vault);
+            config.token(currentClientToken);
+            setTokenExpiry(vault);
+        } else {
+            config.token(currentClientToken);
+        }
+        return vault;
+    }
+
+    protected Vault getVault(VaultConfig config) {
+        return new Vault(config);
+    }
+
+    private void setTokenExpiry(Vault vault) {
+        int tokenTTL = 0;
+        try {
+            tokenTTL = (int) vault.auth().lookupSelf().getTTL();
+        } catch (VaultException e) {
+            LOGGER.log(Level.WARNING, "Could not determine token expiration. " +
+                "Check if token is allowed to access auth/token/lookup-self. " +
+                "Assuming token TTL expired.", e);
+        }
+        tokenExpiry = Calendar.getInstance();
+        tokenExpiry.add(Calendar.SECOND, tokenTTL);
+    }
```

Comment:
Could we not reuse the JCasC package or merge the two?

Reply:
We already have duplicate code in `jcasc.secrets` and `credentials` package.
eg: expiration logic, kubernetes auth logic. 
when we add aws, gcp support for jcacs.secrets, then that might also get duplicated.

We need to create a common package for this two packages to reuse it. It might a bigger refactoring, so can we do it as part of different PR ?

Reply:
@jetersen any update on this ? If you think, we can also try to reuse the JCasC package and merge things

Reply:
I have no strong opinion, my time/interest in the Jenkins project is greatly reduced as I no longer actively use Jenkins 😓

Reply:
Thanks for quick input @jetersen ... We are using this patched version in multiple jenkins. Can we merge this now and create a new release if changes in this PR is good enough ?  We (myself and @ash1425) will make sure that we create another PR to merge JCasC and reduce code duplication.

Reply:
If you want, you could be come a co-maintainer and merge and create the release yourself 😜

Reply:
It is our pleasure to become a co-maintainer to this project. Thanks @jetersen ... We might need your support for the first release creation.

Reply:
It's rather simple process:
Start by requesting release permission here: https://github.com/jenkins-infra/repository-permissions-updater
The readme documents the process.

Once the permission updater has run and you have added your credentials from https://repo.jenkins-ci.org/ to your `~/.m2/settings.xml` your good to go for running a release with `mvn release:prepare release:perform`

Once released and tag is created you can head here: https://github.com/jenkinsci/hashicorp-vault-plugin/releases to use the draft created by release drafter as way to submit the changelog and properly tag the release on GitHub.

Reply:
Okay sure... It is easy process.. Will request the release permission right away. Thanks

Reply:
One small help @jetersen ... In Prerequisite, https://github.com/jenkins-infra/repository-permissions-updater#requesting-permissions it looks I need to have account in https://repo.jenkins-ci.org/webapp/#/home .. Not sure how to open account here. Could you please help here

Reply:
@dineshba https://accounts.jenkins.io/

Reply:
I got access to merge this PR. thanks @jetersen ... I have reviewed this PR today also. I will merge this PR. cc: @ash1425

Reply:
Feel free to go ahead 👏🏼

Reply:
Thanks @jetersen ... Able to create new release `3.7.0`: https://github.com/jenkinsci/hashicorp-vault-plugin/releases and pushed new version to https://repo.jenkins-ci.org/releases/com/datapipe/jenkins/plugins/hashicorp-vault-plugin/ using the command you specified. It was very smooth onboarding. Thanks